northd/northd.c

```diff
@@ -8644,29 +8644,42 @@ build_lswitch_arp_nd_responder_known_ips(struct ovn_port *op,
                 }
             }
         }
-
-        if (op->peer) {
-            const char *arp_proxy = smap_get(&op->nbsp->options,"arp_proxy");
-
+        const char *arp_proxy = smap_get(&op->nbsp->options,"arp_proxy");
+        if (arp_proxy) {
             struct lport_addresses proxy_arp_addrs;
-            int i = 0;
+            int i, ofs = 0;
+            /* Either takes "MAC IP1 IP2" or "IP1 IP2" */
+            if (!extract_addresses(arp_proxy, &proxy_arp_addrs, &ofs) &&
+                !extract_ip_addresses(arp_proxy, &proxy_arp_addrs)) {
+                static struct vlog_rate_limit rl = VLOG_RATE_LIMIT_INIT(1, 5);
+                VLOG_WARN_RL(&rl, "Invalid arp_proxy option: '%s' at lsp '%s'",
+                             arp_proxy, op->nbsp->name);
+                return;
+            }
+
+            /* Select the mac address to answer the proxy ARP/NDP */
+            char* ea_s = NULL;
+            if (!eth_addr_is_zero(proxy_arp_addrs.ea)) {
+                ea_s =  proxy_arp_addrs.ea_s;
+            } else if (op->peer) {
+                ea_s =  op->peer->lrp_networks.ea_s;
+            } else {
+                return;
+            }
 
-            /* Add responses for ARP proxies. */
-            if (arp_proxy && extract_ip_addresses(arp_proxy,
-                                                  &proxy_arp_addrs) &&
-                proxy_arp_addrs.n_ipv4_addrs) {
+            /* Add IPv4 responses for ARP proxies. */
+            if (proxy_arp_addrs.n_ipv4_addrs) {
                 /* Match rule on all proxy ARP IPs. */
                 ds_clear(match);
                 ds_put_cstr(match, "arp.op == 1 && arp.tpa == {");
 
                 for (i = 0; i < proxy_arp_addrs.n_ipv4_addrs; i++) {
-                    ds_put_format(match, "%s,",
-                                  proxy_arp_addrs.ipv4_addrs[i].addr_s);
+                    ds_put_format(match, "%s/%u,",
+                                  proxy_arp_addrs.ipv4_addrs[i].addr_s, proxy_arp_addrs.ipv4_addrs[i].plen);
```

Comment:
`proxy_arp_addrs.ipv4_addrs[i].plen` should be on new line aligned with `match`.